text/3637-guard-patterns.md

```diff
@@ -0,0 +1,250 @@
+        // complete the transaction
+    }
+    _ => {
+        // the user doesn't have enough credit, return an error message
+    }
+}
+```
+
+Now we have just one arm for a successful transaction, with an or-pattern combining the two arms we used to have. The two nested patterns are of the form
+
+```rust
+pattern if expr
+```
+
+This is a **guard pattern**. It matches a value if `pattern` (the pattern it wraps) matches that value, _and_ `expr` evaluates to `true`. Like in match arm guards, `expr` can use values bound in `pattern`.
+
+## For New Users
+
+For new users, guard patterns are better explained without reference to match arm guards. Instead, they can be explained by similar examples to the ones currently used for match arm guards, followed by an example showing that they can be nested within other patterns and used outside of match arms.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Supersession of Match Arm Guards
+
+Rather than being parsed as part of the match expression, guards in match arms will instead be parsed as a guard pattern. For this reason, the `if` pattern operator must have lower precedence than all other pattern operators.
+
+That is,
+
+```rs
+// Let <=> denote equivalence of patterns.
+
+x @ A(..) if pred       <=> (x @ A(..)) if pred
+&A(..) if pred          <=> (&A(..)) if pred
+A(..) | B(..) if pred   <=> (A(..) | B(..)) if pred
+```
+
+## Interaction with Expression Operators
+
+The or-pattern operator and the bitwise OR operator both use the `|` token. This creates a parsing ambiguity:
+
+```rust
+// How is this parsed?
+false if foo | true 
+// As a guard pattern nested in an or-pattern?
+(false if foo) | true
+// Or as a pattern guarded by a bitwise OR operation?
+false if (foo | true)
+```
+
+For that reason, guard patterns nested within or-patterns must be explicitly parenthesized.
+
+There's a similar ambiguity between `=` used as the assignment operator within the guard
+and used outside to indicate assignment to the pattern (e.g. in `if`-`let`, `while let`, etc.).
+Therefore guard patterns appearing at the top level in those places must also be parenthesized:
+
+```rust
+while let x if guard(x) = foo() {}   // not allowed
+while let (x if guard(x)) = foo() {} // allowed
+```
+
+Therefore the syntax for patterns becomes
+
+> **<sup>Syntax</sup>**\
+> _Pattern_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; _PatternNoTopGuard_\
+> &nbsp;&nbsp; | _GuardPattern_
+> 
+> _PatternNoTopGuard_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; `|`<sup>?</sup> _PatternNoTopAlt_  ( `|` _PatternNoTopAlt_ )<sup>\*</sup>
+
+With `if let`, `while let`, and `for` expressions now using `PatternNoTopGuard`. `let` statements can continue to use `PatternNoTopAlt`.
+
+## Bindings Available to Guards
+
+The only bindings available to guard conditions are
+- bindings from the scope containing the pattern match, if any; and
+- bindings introduced by identifier patterns _within_ the guard pattern.
```

Comment:
with #2294 (rust-lang/rust#51114) the guard can be an `if let` clause that introduces binding to the match arm:

```rust
#![feature(if_let_guard)]

fn main() {
    match Some(Some(4)) {
        Some(a) if let Some(b) = a => assert_eq!(b, 4),
        _ => unreachable!(),
    }
}
```

how will if-let guard interact with this RFC?

```rust
// error?
(Some(a) if let Some(b) = a.foo()) => using(b),

// error?
Some(a if let Some(b) = a.foo()) => using(b),

// error? assuming the `b` on both branches have the same type
(Some(Some(a)) if let Ok(b) = a.foo()) | (Some(Some(a)) if let Err(b) = a.bar()) => using(b),

// definitely error
(Some(a) if let Some(b) = a.foo()) | None => using(b),
```

Reply:
Hmm, I wasn't aware of this feature. If you look in the Future Possibilities section I do mention the possibility of allowing `if let`, but I hadn't realized it would be necessary for this RFC. I think probably all three of those patterns should be allowed, but to be cautious it might be better to restrict it to work only on the top-level for now.

Reply:
It's not necessary since if let guards aren't stable. Most likely when both features are enabled together then `if let` guards would be allowed in patterns everywhere that `if` guards are. Implementation-wise we can also forbid combining these two features to start with if that proves too tricky.